### DISCUSSION TITLE
How to use single database connection throughout application.

### DISCUSSION THREAD

**Discussion by Harsh-Kesharwani** [fastapi/fastapi#11625] [Questions] [answered]
### Privileged issue

- [ ] I'm @tiangolo or he asked me directly to create an issue here.

### Issue Content

How to use single database connection throughout application using sqlalchemy. I have to interact with two different database, i do not want to create database session  on every request. is there any way to create single database connection for each database when application started.

**Comment by YuriiMotov** [comment]
Why do you want to use one single connection per database?
You should keep in mind that requests will be processed not sequentially, they may be processed in parallel and this may brake your app if you use the same session in different requests (https://docs.sqlalchemy.org/en/20/orm/session_basics.html#is-the-session-thread-safe-is-asyncsession-safe-to-share-in-concurrent-tasks).

It is better to use different sessions and get them using the dependency mechanism.
If you worry about overheads (establishing a connection takes time), SQLAlchemy can reuse connections. You can limit the amount of connections in connection pool as well by specifying `pool_size=1, max_overflow=0` as a parameters of `create_engine` (or `create_async_engine` if you use async connection).

Read more about connection pooling: https://docs.sqlalchemy.org/en/20/core/pooling.html

And if you still want to create one session per DB, you can create it in the lifespan event and store into `state`. Then you can access it using `request.state`:

```
# I advise against using this sample code. It's just for demonstration

from contextlib import asynccontextmanager

from fastapi import FastAPI, Request

@asynccontextmanager
async def lifespan(app: FastAPI):
    db_connection_1 = Session()
    yield {"db_connection_1": db_connection_1}

app = FastAPI(lifespan=lifespan)

@app.get("/")
def index(request: Request):
    db_connection_1 = request.state.db_connection_1
    pass
```

**Comment by PhysicallyActive** [comment]
If you want two different sqlalchemy sessionmakers, you can just create two different ones?

```python
from sqlalchemy import create_engine
from sqlalchemy.orm import sessionmaker

CONNECTION_STRING_1 = "sqlite:///./databases/myFirstDB.db"
CONNECTION_STRING_2 = "sqlite:///./databases/mySecondDB.db"

engine_1 = create_engine(
    CONNECTION_STRING_1,
    connect_args={"check_same_thread": False},
)
engine_2 = create_engine(
    CONNECTION_STRING_2,
    connect_args={"check_same_thread": False},
)
session_generator_1 = sessionmaker(autocommit=False, autoflush=False, bind=engine_1)
session_generator_2 = sessionmaker(autocommit=False, autoflush=False, bind=engine_2)
```

From this you can create two dependency injectable functions `async def get_db_1()` and `async def get_db_2()` (as detailed [here](https://fastapi.tiangolo.com/tutorial/dependencies/dependencies-with-yield/#a-database-dependency-with-yield))

@Edit: Just saw @YuriiMotov's answer. I might have mistunderstood the question.

**Comment by Harsh-Kesharwani** [comment]
@YuriiMotov @PhysicallyActive Thanks. 
i want one help, does below code creates database connection pool on every request?  due to this code i have established too many connections and my database stops working. also i am using context var so that i can use this database connection in nested function calls without explicitly passing them. 

```
def x_database():
    DB_HOST = os.getenv("DB_X_HOST")
    DB_USER = os.getenv("DB_X_USER")
    DB_PASSWORD = os.getenv("DB_X_PASSWORD")
    DB_NAME = os.getenv("DB_X_DATABASE")

    escaped_password = urllib.parse.quote_plus(DB_PASSWORD)

    DATABASE_URL = f"mysql+pymysql://{DB_USER}:{escaped_password}@{DB_HOST}/{DB_NAME}"

    engine = create_engine(DATABASE_URL)

    SessionLocal = sessionmaker(autocommit=False, autoflush=False, bind=engine)

    return SessionLocal()
```

```
def get_db_x():
    db = x_database()
    try:
        yield db
    finally:
        db.close()
```

```
from contextvars import ContextVar
from sqlalchemy.orm import Session

db_session_x: ContextVar[Session] = ContextVar('db_session')
```
```
@router.get("/testing")
def testing(db_x: Annotated[Session, Depends(get_db_x)]):
    db_session_x.set(db_x)

    db_xy = db_session_x.get()
    data = db_xy.query(Airport).filter(Airport.id == 1).first()

    return data
```

**Reply by PhysicallyActive** [reply]
With this approach you'll create a new DB-connection (via a new engine) every time someone calls your `/testing` endpoint.
Instead, you'd benefit from creating either a global-scoped `sessionmaker` (similar to my previous post) or a singleton class that generates `Session`s when a function is called.

**Reply by YuriiMotov** [reply]
Yes, your `x_database` will be called on every request and it creates new connection pool every time.
You need to execute the code from `x_database` only once on app startup and store `SessionLocal` object somewhere, then use this stored `SessionLocal` in `get_db_x` to get a session.

Another way is to cache results of `x_database` function by adding a `functools.cache` decorator to it:
```
@functools.cache
def x_database():
    ...
```

**Reply by Harsh-Kesharwani** [reply]
@PhysicallyActive @YuriiMotov Thanks. I think i should not  use SessionLocal() inside a function and also i can use cache for safer side. i can simply change my config/database.py file to below one. this is correct?
Also can you tell me how to invalidate cache or what is default timeout for cache invalidation.
```
    DB_HOST = os.getenv("DB_X_HOST")
    DB_USER = os.getenv("DB_X_USER")
    DB_PASSWORD = os.getenv("DB_X_PASSWORD")
    DB_NAME = os.getenv("DB_X_DATABASE")

    escaped_password = urllib.parse.quote_plus(DB_PASSWORD)

    DATABASE_URL = f"mysql+pymysql://{DB_USER}:{escaped_password}@{DB_HOST}/{DB_NAME}"

    engine = create_engine(DATABASE_URL)

    SessionLocal = sessionmaker(autocommit=False, autoflush=False, bind=engine)

@functools.cache
def get_db_x():
    db = SessionLocal()
    try:
        yield db
    finally:
        db.close()
 ```
 
 Now i can use this get_db_x inside router file using dependency injection. hope now this will not create connection pool on each request and it will be global to all request.

**Reply by YuriiMotov** [reply]
No, you shouldn't decorate `get_db_x` with `@functools.cache`. This will not work.
You can either cache the results of `x_database` *using `cache` decorator) or move code of `x_database` outside the function.

**Reply by YuriiMotov** [reply]
And, I think you don't need to invalidate cache in this use case

**Reply by Harsh-Kesharwani** [reply]
@YuriiMotov you mean below code is correct. can you please me how to use db dependency in nested function calls. currently i am using context var for this. I have same issue as mention in this [issue](https://github.com/tiangolo/fastapi/discussions/6889 )

```
    DB_HOST = os.getenv("DB_X_HOST")
    DB_USER = os.getenv("DB_X_USER")
    DB_PASSWORD = os.getenv("DB_X_PASSWORD")
    DB_NAME = os.getenv("DB_X_DATABASE")

    escaped_password = urllib.parse.quote_plus(DB_PASSWORD)

    DATABASE_URL = f"mysql+pymysql://{DB_USER}:{escaped_password}@{DB_HOST}/{DB_NAME}"

    engine = create_engine(DATABASE_URL)

    SessionLocal = sessionmaker(autocommit=False, autoflush=False, bind=engine)

def get_db_x():
    db = SessionLocal()
    try:
        yield db
    finally:
        db.close()
```

**Reply by YuriiMotov** [reply]
I think that the best way is to pass session as a parameter to these nested functions

**Reply by Harsh-Kesharwani** [reply]
@YuriiMotov You are correct, but some function may not required session and other nested function may required session. so always passing session as a parameter will be too hectic. can i create middle where and use context var to store session like below given:

```
@app.middleware("http")
async def store_context(request: Request, call_next, db_x: Annotated[Session, Depends(get_db_x)]):
    db_session_x.set(db_x)
    try:
        payload_data = await request.json()
    except json.JSONDecodeError:
        payload_data = {}

    query_params = dict(request.query_params)

    data = {
        'payload_data': payload_data,
        'query_params': query_params
    }
    request_data_vars.set(data)
    
    response = await call_next(request)
    return response
``` 

context_var.py
```
from contextvars import ContextVar
from sqlalchemy.orm import Session

db_session_x: ContextVar[Session] = ContextVar('db_session')
db_session_y: ContextVar[Session] = ContextVar('db_session')

request_data_vars = ContextVar("request_data")
```

finally i can use like this in router file:
```
@router.get("/testing")
def testing():
    db_x = db_session_x.get()
    data = db_x.query(Airport).filter(Airport.id == 1).first()

    return data
```

Is this good practice to do ? , or should i follow other approach of doing this.

**Reply by YuriiMotov** [reply, marked as answer]
You can't use `Depends` in middleware.. It's only for endpoint function parameters.
I think you can set context var in your endpoint function like it's shown in this answer: https://github.com/tiangolo/fastapi/discussions/6889#discussioncomment-5132510

**Comment by Harsh-Kesharwani** [comment]
@YuriiMotov thanks

**Reply by iiotsrc** [reply]
Hai good evening 
Can I know u posted issues completely solved ah?
